syncer/db.go

```diff
@@ -69,45 +69,85 @@ type binlogSize struct {
 	size int64
 }
 
-// Conn represents a live DB connection
-type Conn struct {
-	cfg      *config.SubTaskConfig
-	baseConn *baseconn.BaseConn
+func createBaseDB(dbCfg config.DBConfig) (*conn.BaseDB, error) {
+	db, err := conn.DefaultDBProvider.Apply(dbCfg)
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
 }
 
-// ResetConn reset baseConn.*DB's connection pool
-func (conn *Conn) ResetConn(tctx *tcontext.Context) error {
-	if conn.baseConn == nil {
-		return terror.ErrDBUnExpect.Generate("database base connection not valid")
-	}
-	return conn.baseConn.ResetConn(tctx)
+// UpStreamConn connect to upstream DB, use *sql.DB instead of *sql.Conn
+type UpStreamConn struct {
+	cfg    *config.SubTaskConfig
+	BaseDB *conn.BaseDB
 }
 
-func (conn *Conn) getMasterStatus(flavor string) (mysql.Position, gtid.Set, error) {
-	return utils.GetMasterStatus(conn.baseConn.DB, flavor)
+func (conn *UpStreamConn) getMasterStatus(flavor string) (mysql.Position, gtid.Set, error) {
+	return utils.GetMasterStatus(conn.BaseDB.DB, flavor)
 }
 
-func (conn *Conn) getServerUUID(flavor string) (string, error) {
-	return utils.GetServerUUID(conn.baseConn.DB, flavor)
+func (conn *UpStreamConn) getServerUUID(flavor string) (string, error) {
+	return utils.GetServerUUID(conn.BaseDB.DB, flavor)
 }
 
-func (conn *Conn) getParser(ansiQuotesMode bool) (*parser.Parser, error) {
-	return utils.GetParser(conn.baseConn.DB, ansiQuotesMode)
+func (conn *UpStreamConn) getParser(ansiQuotesMode bool) (*parser.Parser, error) {
+	return utils.GetParser(conn.BaseDB.DB, ansiQuotesMode)
 }
 
-func (conn *Conn) killConn(connID uint32) error {
-	return utils.KillConn(conn.baseConn.DB, connID)
+func (conn *UpStreamConn) killConn(connID uint32) error {
+	return utils.KillConn(conn.BaseDB.DB, connID)
 }
 
-func (conn *Conn) fetchAllDoTables(bw *filter.Filter) (map[string][]string, error) {
-	return utils.FetchAllDoTables(conn.baseConn.DB, bw)
+func (conn *UpStreamConn) fetchAllDoTables(bw *filter.Filter) (map[string][]string, error) {
+	return utils.FetchAllDoTables(conn.BaseDB.DB, bw)
 }
 
-func (conn *Conn) countBinaryLogsSize(pos mysql.Position) (int64, error) {
-	return countBinaryLogsSize(pos, conn.baseConn.DB)
+func (conn *UpStreamConn) countBinaryLogsSize(pos mysql.Position) (int64, error) {
+	return countBinaryLogsSize(pos, conn.BaseDB.DB)
 }
 
-func (conn *Conn) querySQL(tctx *tcontext.Context, query string, args ...interface{}) (*sql.Rows, error) {
+func createUpStreamConn(cfg *config.SubTaskConfig, dbCfg config.DBConfig) (*UpStreamConn, error) {
+	baseDB, err := createBaseDB(dbCfg)
+	if err != nil {
+		return nil, terror.WithScope(terror.DBErrorAdapt(err, terror.ErrDBDriverError), terror.ScopeUpstream)
+	}
+	return &UpStreamConn{BaseDB: baseDB, cfg: cfg}, nil
+}
+
+func closeUpstreamConn(tctx *tcontext.Context, conn *UpStreamConn) {
+	if conn != nil {
```

Comment:
and can we reuse function `closeBaseDB`?

Reply:
yes, we can reuse function `closeBaseDB`, I fixed it and it will check nil [inside](https://github.com/pingcap/dm/pull/266/files#diff-c3cc6fccd12f8e1679a575d1aaf8fd14R81)